pkg/apiserver/rest/dualwriter_mode1.go

```diff
@@ -49,15 +52,20 @@ func (d *DualWriterMode1) Create(ctx context.Context, original runtime.Object, c
 	}
 	d.recordLegacyDuration(false, mode1Str, options.Kind, method, startLegacy)
 
+	d.pendingActions.Add(1)
 	go func() {
+		defer d.pendingActions.Done()
+		ctx := context.WithoutCancel(ctx)
```

Comment:
If we don't use `WithoutCancel`, the context can either be canceled because of the timeout we set below with `WithTimeoutCause` or because the parent was canceled, which we don't want.

Reply:
> or because the parent was canceled, which we don't want.

are you _sure?_ it seems like calling `Create` with a cancelled context is a very bad idea that will lead to bugs...

Reply:
The parent context may be canceled at any time. For instance, once the request is served the server may opt to cancel it. Consider the following example server:
```go
func (s *server) serve(ctx context.Context, l net.Listener) error {
    ctx, cancel := context.WithCancel(ctx)
    defer cancel()
    for {
        conn, err := l.Accept()
        // handle err
        go s.handleConn(ctx, conn)
    }
    return nil
}

func (s *server) handleConn(ctx context.Context, conn *net.TCPConn) {
    ctx, cancel := context.WithCancel(ctx)
    defer cancel()
    req, err := parseRequest(conn)
    // handle err
    s.handleRequest(ctx, req)
}
```
This means that as soon as the request is served, the context derived and passed to the handler is canceled. This ensures that all handlers are bound by the server context, and none of them leak. In reality, the server typically does more work after serving the request and before canceling the derived context passed to the handler, and the goroutine we created to make the background `Create` will fail doing any context-aware work, like writing to a database. While it is true that using `context.WithoutCancel` allows to circumventing the server providing the transport operation to prevent handler leaks, we provide our own server that is allowed to do work outside of the transport server (i.e., the same as a server that serves over HTTP but also has "cron"-like jobs that are non-HTTP).

The concern about values in the context that may be tied to a single request, and thus may produce unpredictable behaviour if the context is canceled is valid. For instance, if an observability decorator flushes to the network the spans created during the execution of the request and then releases some pointers, a subsequent method call to create a new span may panic. But how do we track the work that needs to branch from a request, like something that needs retrying in the background, etc.? This lead to this [discussion](https://github.com/golang/go/issues/40221) and the creation of `context.WithoutCancel`, and the recommendation is that decorators are aware and gracefully handle these conditions. For example, in OpenTracing, we can create a span with `FollowsFrom` that allows to represent this detachment from a request (I don't recall the name for OpenTelementry, but it's similar).

So yeah, we need to account for branching operations with context in our decorators.

Reply:
I'm aware of the usefulness of `context.WithoutCancel`, but it needs to be used sparingly and only in cases where it's safe to do so.

Maybe I should rephrase my question:

Why is it desirable to continue creating resources if the context has been cancelled?

A potential source of data integrity bugs involves performing operations while the server is shutting down. If the context has been cancelled because the server is shutting down and we blindly ignore that cancellation and continue trying to do I/O, we may end up persisting inconsistent or corrupted data.

Reply:
I would agree that we would want to cancel if the parent context got cancelled. In mode1, the only thing we need to guarantee is that we write to legacy store. Writing to unified store works as a best effort. And actually, one thing we want to avoid is succeeding writing to unified storage when _for some reason_ writing to legacy fails (because the context was cancelled, for example).

Reply:
> Why is it desirable to continue creating resources if the context has been cancelled?

I think the best way to phrase the general intention would be: try to do some work without blocking the response to the user to have the least impact on them. But past the moment of returning from a handler makes it very likely that the context is cancelled very shortly, and the background goroutine has just started and it still needs to make some I/O *(i.e. call the db), so it's very likely that it will just become a noisy nop just failing most of the time.

> A potential source of data integrity bugs involves performing operations while the server is shutting down.

Totally agree. That's why I wrote [this](https://github.com/grafana/grafana/pull/88886/files#r1631525475) comment mentioning that we should consider server lifecycle, and the comments on the code mention that this is a temporary workaround that needs to be fixed since past the execution of the main loop the status of the storage is undefined.

Reply:
But I do agree with you that introducing `WihtoutCancel` here without adding the correct mechanism for handling the server lifecycle would in itself be introducing a bug. So I think this should be taken away, thanks for pointing that out!